site-src/geps/gep-726.md

```diff
@@ -0,0 +1,319 @@
+
+## Goals
+
+* Implement path redirects.
+* Implement the most portable and simple forms of path rewrites.
+* Describe how more advanced rewrite and redirect and redirect capabilities
+  could be added in the future.
+
+## API
+
+Although many implementations support very advanced rewrite and redirect
+capabilities, the following are the most [portable](#portability) concepts that
+are not already supported by the Gateway API:
+
+* Path redirects
+* Path prefix redirects
+* Path prefix rewrites
+* Host rewrites
+
+Although regular expression based redirects and rewrites are commonly supported,
+there is significantly more variation in both if and how they are implemented.
+Given the wide support for this concept, it is important to design the API in a
+way that would make it easy to add this capability in the future.
+
+### Path Modifiers
+
+Both redirects and rewrites would share the same `PathModifier` types:
+
+```go
+// HTTPPathModifierType defines the type of path redirect.
+type HTTPPathModifierType string
+
+const (
+  // This type of modifier indicates that the complete path will be replaced by
+  // the path redirect value.
+  AbsoluteHTTPPathModifier HTTPPathModifierType = "Absolute"
+
+  // This type of modifier indicates that any prefix path matches will be
+  // replaced by the substitution value. For example, a path with a prefix match
+  // of "/foo" and a ReplacePrefixPathMatch substitution of "/bar" will have the
+  // "/foo" prefix replaced with "/bar" in matching requests.
+  PathMatchHTTPPathModifier HTTPPathModifierType = "ReplacePrefixPathMatch"
+)
+
+// HTTPPathModifier defines configuration for path modifiers.
+type HTTPPathModifier struct {
+  // Type defines the type of path modifier.
+  //
+  // +kubebuilder:validation:Enum=Absolute;ReplacePrefixPathMatch
+  Type HTTPPathModifierType `json:"type"`
+
+  // Substitution defines the HTTP path value to substitute. An empty value ("")
+  // indicates that the portion of the path to be changed should be removed from
+  // the resulting path. For example, a request to "/foo/bar" with a prefix
+  // match of "/foo" would be modified to "/bar".
+  //
+  // +kubebuilder:validation:MaxLength=1024
+  Substitution string `json:"substitution"`
+}
+```
+
+### Redirects
+
+The existing `RequestRedirect` filter can be expanded to support path redirects.
+In the following example, a request to `/foo/abc` would be redirected to
+`/bar/abc`.
+
+```yaml
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1alpha2
+metadata:
+  name: http-filter-1
+spec:
+  rules:
+    - matches:
+      - path:
+          type: Prefix
+          value: /foo
+      filters:
+      - type: RequestRedirect
+        requestRedirect:
+          hostname: foo.com
+          path:
+            type: ReplacePrefixPathMatch
+            value: /bar
+```
+
+This would be represented with the following API additions:
+```go
+// HTTPRequestRedirect defines a filter that redirects a request. At most one of
```

Comment:
We should state very explicitly this sends a 3xx response to the client.

Reply:
This part of the GEP is confusing because it was not clear that the redirect filter itself already exists and is more clearly defined than this snippet suggests. I've updated the GEP to clarify that this is an addition to an existing filter. https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha2/httproute_types.go#L703